### PR TITLE
E98 bird

### PR DESCRIPTION
**Pull request by dudiq**

### CODE REVIEW

**Comment by dudiq** [comment]
@kazurus approved

**Comment by dudiq** [comment]
@lamartire ready for review

File: src/class/singleton/signer/Wallet.js [hunk 1]
```diff
@@ -128,28 +129,22 @@ export default class Wallet {
...
-        resolve(trxHash);
-      });
-      sendEvent.on('error', error => {
-        reject(error);
-      });
-      sendEvent.catch(error => {
-        reject(error);
-      });
-    });
+    const hash = await web3.call('eth_sendRawTransaction', signedTx);
```
Comment (kazurus) [inline src/class/singleton/signer/Wallet.js]: remove raw calls

File: src/class/singleton/signer/proxySigner.js [hunk 2]
```diff
@@ -29,8 +29,4 @@ export default {
     const Singer = await getSigner();
     return Singer.setWeb3Network(network);
   },
-  async getWeb3Instance() {
```
Comment (kazurus) [inline src/class/singleton/signer/proxySigner.js]: can we lazy load web3?

File: src/components/screens/DocumentCreate/DocumentCreateController.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 // @ts-check
 import { VuexModule, Action, Module } from 'vuex-class-modules';
-import ConnectError from '@endpass/connect/ConnectError';
+import ConnectError from '@endpass/connect/connect-error';
```
Comment (kazurus) [inline src/components/screens/DocumentCreate/DocumentCreateController.js]: we should show that connect error is class

File: src/components/widget/Header.vue [hunk 4]
```diff
@@ -108,7 +104,12 @@ export default {
     },
 
     actualBalance() {
-      const balanceInEth = fromWei(this.balance);
+      // TODO: move this logic to store
+      if (this.$options.accountsStore.isBalanceLoading) {
```
Comment (kazurus) [inline src/components/widget/Header.vue]: remove

File: src/components/widget/Header.vue [hunk 5]
```diff
@@ -141,9 +142,13 @@ export default {
     },
 
     isLoading() {
-      const { balance, isBalanceInFiat, ethPriceInFiat } = this;
+      if (this.$options.accountsStore.isBalanceLoading) {
```
Comment (kazurus) [inline src/components/widget/Header.vue]: pass balance and loading

File: src/class/singleton/signer/signer.js [hunk 6]
```diff
@@ -67,8 +63,9 @@ export default {
     }
   },
 
-  setWeb3Network,
-  getWeb3Instance() {
-    return web3;
+  setWeb3Network(netId) {
+    const nets = Network.NETWORK_URL[netId];
```
Comment (lamartire) [inline src/class/singleton/signer/signer.js]: `const [netUrl] = Network.NETWORK_URL[netId]`

File: src/store/modules/AccountsModule.js [hunk 7]
```diff
@@ -41,15 +49,56 @@ class AccountsModule extends VuexModule {
     return !!this.settings.otpEnabled;
   }
 
+  get ethBalance() {
+    if (!this.isBalanceLoading && this.balance) {
```
Comment (kazurus) [inline src/store/modules/AccountsModule.js]: change order

File: src/store/modules/AccountsModule.js [hunk 8]
```diff
@@ -41,15 +49,56 @@ class AccountsModule extends VuexModule {
...
+    if (isNetworkChanged) {
+      setWeb3Network(newSettings.net);
+    }
+
+    this.resubscribeBalance({ isNetworkChanged, newSettings, oldSettings });
+  }
+
+  @Action
+  async resubscribeBalance({ isNetworkChanged, newSettings, oldSettings }) {
+    if (!this.isBalanceAutoUpdate) {
```
Comment (kazurus) [inline src/store/modules/AccountsModule.js]: we can do better :smiley:

File: src/class/singleton/web3.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+const web3 = new Web3Api({
+  netUrl: defaultUrl,
+});
+
+export const setWeb3Network = (netId = Network.NET_ID.MAIN) => {
+  // :TODO change isMock checking
+  if (
+    web3.currentProvider &&
+    web3.currentProvider.connection &&
+    web3.currentProvider.connection.isMock
```
Comment (lamartire) [inline src/class/singleton/web3.js]: May be just use `lodash/get`?

File: src/store/modules/AccountsModule.js [hunk 10]
```diff
@@ -41,15 +49,56 @@ class AccountsModule extends VuexModule {
...
+  async changeSettings(newSettings) {
+    const oldSettings = this.settings;
+    this.settings = newSettings;
+
+    const isNetworkChanged = oldSettings.net !== newSettings.net;
+    if (isNetworkChanged) {
+      setWeb3Network(newSettings.net);
+    }
+
+    this.resubscribeBalance({ isNetworkChanged, newSettings, oldSettings });
```
Comment (lamartire) [inline src/store/modules/AccountsModule.js]: Leave `TODO` here about resubscribing to `eth` prices in future, when related logic will appear here

File: src/store/modules/AccountsModule.js [hunk 11]
```diff
@@ -41,15 +49,56 @@ class AccountsModule extends VuexModule {
...
+    if (!this.isBalanceAutoUpdate) {
+      return;
+    }
+
+    const isLastAccountChanged =
+      newSettings.lastActiveAccount &&
+      oldSettings.lastActiveAccount &&
+      oldSettings.lastActiveAccount !== newSettings.lastActiveAccount;
+
+    const isBalanceUpdateStopped = !this.isBalanceLoading && !this.balance;
```
Comment (lamartire) [inline src/store/modules/AccountsModule.js]: So, what about cases when we really have zero balance? Is it really sign of balance updates stopping?

File: src/store/modules/AccountsModule.js [hunk 12]
```diff
@@ -261,34 +305,41 @@ class AccountsModule extends VuexModule {
...
+    const address = get(this.settings, 'lastActiveAccount');
+    const data = await web3.getBalance(address);
 
-    return balance;
+    return data;
+  }
+
+  @Action
+  async enableAutoUpdateBalance() {
+    this.isBalanceAutoUpdate = true;
```
Comment (lamartire) [inline src/store/modules/AccountsModule.js]: I think it is better predicate than `isBalanceUpdateStopped`

File: src/store/modules/AccountsModule.js [hunk 13]
```diff
@@ -261,34 +305,41 @@ class AccountsModule extends VuexModule {
...
+      this.balance = result;
+
+      if (isNetworkChanged) {
+        this.subscribeOnBalanceUpdates();
+        break;
+      }
+
+      if (address !== get(this.settings, 'lastActiveAccount') || error) {
+        break;
+      }
```
Comment (lamartire) [inline src/store/modules/AccountsModule.js]: What about case when we are manually disabling auto updates?

File: src/store/modules/AccountsModule.js [hunk 14]
```diff
@@ -316,10 +367,10 @@ class AccountsModule extends VuexModule {
     return authService.getSeedTemplateUrl();
   }
 
-  @Mutation
+  @Action
   logout() {
     this.accounts = [];
-    this.settings = {};
+    this.changeSettings({});
```
Comment (lamartire) [inline src/store/modules/AccountsModule.js]: I think we also should kill all subscriptions here

File: src/store/modules/RequestsModule.js [hunk 15]
```diff
@@ -21,10 +23,8 @@ class RequestsModule extends VuexModule {
 
   @Action
   async getNextNonce(address) {
-    const web3 = await signer.getWeb3Instance();
-    const nonce = await web3.eth.getTransactionCount(address);
-
-    return nonce;
+    const nonce = await web3.call('eth_getTransactionCount', address);
```
Comment (lamartire) [inline src/store/modules/RequestsModule.js]: So, what do you think, @kazurus, @dudiq and @ifiri, about having "sugar" methods for each method in `web3`? Like `web3.getTransactionCount`? I guess it should not make library bloat as original `web3`.

File: tests/unit/mocks/service/Signer/web3.mock.js [hunk 16]
```diff
@@ -1,18 +1,53 @@
...
-      eth: {
-        sendSignedTransaction: jest.fn(),
-        getCode: jest.fn().mockResolvedValue('0x'),
-      },
-      providers: {
-        HttpProvider: jest.fn(),
-      },
-      setProvider: jest.fn(),
-    },
+// jest.mock('@/class/singleton/signer/web3', () => {
```
Comment (lamartire) [inline tests/unit/mocks/service/Signer/web3.mock.js]: Remove that

File: tests/unit/mocks/service/Signer/web3.mock.js [hunk 17]
```diff
@@ -1,18 +1,53 @@
...
+  asyncIteratorMock[Symbol.asyncIterator] = async function*() {
+    yield { result: '100000000000000' };
   };
+  p.iterateBalance = jest.fn().mockImplementation(() => {
+    return asyncIteratorMock;
+  });
+
+  return PublicApi;
+  //
+  // return {
```
Comment (lamartire) [inline tests/unit/mocks/service/Signer/web3.mock.js]: This too

File: tests/unit/mocks/web3.mock.js [hunk 18]
```diff
@@ -1,19 +1,19 @@
...
-    eth,
-  }));
-  // Allows you to replace stubs of web3 instance methods in unit tests
-  mockWeb3.eth = eth;
-
-  return mockWeb3;
-});
-
-export default Web3;
+// import Web3 from 'web3';
```
Comment (lamartire) [inline tests/unit/mocks/web3.mock.js]: Just delete unused code

File: vue.config.js [hunk 19]
```diff
@@ -44,7 +45,7 @@ module.exports = {
   configureWebpack: {
     devtool: SOURCE_MAP && 'cheap-module-eval-source-map',
     resolve: {
-      symlinks: false,
+      // symlinks: false,
```
Comment (lamartire) [inline vue.config.js]: 💥

File: src/components/widget/Header.vue (hunk 4 above)
Comment (lamartire) [inline src/components/widget/Header.vue]: I think on loading moment we can't have any balance

**Comment by lamartire** [comment]
Can't test it locally because have a lot of unexpected issues with auth and can't sign in for testing inner features

File: src/store/modules/RequestsModule.js (hunk 15 above)
Comment (kazurus) [reply src/store/modules/RequestsModule.js]: We have already discussed, will add

**Pull request merged**